### PR TITLE
SerdeAnyMap: add `unsafe_stable_anymap` feature that uses `type_name` instead of `TypeId::of`

### PR DESCRIPTION
**Pull request by domenukk**
While _technically_ unsound, this just might give us a build-agnostic anymap which would pre pretty nice to have to store and load state.

### CODE REVIEW

**Comment by domenukk** [comment]
@addisoncrump just _how_ stupid of an idea is this?

**Comment by addisoncrump** [comment]
Since we already require `'static`, it's actually safe afaik. That said, it will be significantly more expensive to compute since we end up hashing what could be a very long name to enter the item into the hashmap.

**Comment by addisoncrump** [comment]
Ah, actually, it theoretically could be problematic if the type names (not including the crate path) are exactly the same, but I'm quite certain this won't happen in practice.

**Comment by domenukk** [comment]
If they are all `'static` we could also `&'static str` directly? That sounds faster to compare in the best case?

**Comment by domenukk** [comment]
Sad, doesn't seem to work:
```rust
error: lifetime may not live long enough
   --> libafl_bolts/src/serdeany.rs:143:21
    |
132 |     impl<'de> serde::de::Visitor<'de> for BoxDynVisitor {
    |          --- lifetime `'de` defined here
...
143 |             let id: TypeRepr = visitor.next_element()?.unwrap();
    |                     ^^^^^^^^ type annotation requires that `'de` must outlive `'static`

error: lifetime may not live long enough
   --> libafl_bolts/src/serdeany.rs:227:9
    |
225 |     #[derive(Debug, Serialize, Deserialize)]
    |                                ----------- lifetime `'de` defined here
226 |     pub struct SerdeAnyMap {
227 |         map: HashMap<TypeRepr, Box<dyn SerdeAny>>,
    |         ^^^ requires that `'de` must outlive `'static`

error: lifetime may not live long enough
   --> libafl_bolts/src/serdeany.rs:424:9
    |
422 |     #[derive(Debug, Serialize, Deserialize)]
    |                                ----------- lifetime `'de` defined here
423 |     pub struct NamedSerdeAnyMap {
424 |         map: HashMap<TypeRepr, HashMap<u64, Box<dyn crate::serdeany::SerdeAny>>>,
    |         ^^^ requires that `'de` must outlive `'static`

```

**Comment by addisoncrump** [comment]
We can manually impl Deserialize here.

**Comment by addisoncrump** [comment]
> If they are all `'static` we could also `&'static str` directly? That sounds faster to compare in the best case?

Static lifetime means that, if I own the value, it will live for as long as I hold it. It doesn't guarantee that these values will be in the same place between executions, so we can't serialize them as &'static str. Instead, we'll need to manually deserialise to match the entries in the local registry.

**Comment by addisoncrump** [comment]
Actually, we can do a sneaky and derive Deserialize for a wrapper type around the TypeRepr to simplify this.

**Comment by domenukk** [comment]
Do you want to give it a shot?

**Comment by addisoncrump** [comment]
Can you commit the non-working code? I'll fix it from there.

**Comment by domenukk** [comment]
I just replaced String with `&'static str` as `TypeRepr`

File: libafl_bolts/src/serdeany.rs [hunk 1]
```diff
@@ -98,7 +140,7 @@ pub mod serdeany_registry {
         where
             V: serde::de::SeqAccess<'de>,
         {
-            let id: u128 = visitor.next_element()?.unwrap();
+            let id: TypeRepr = visitor.next_element()?.unwrap();
```
Comment (domenukk) [inline libafl_bolts/src/serdeany.rs:143]: @addisoncrump I'm not 100% sure if this type annotation is correct (or should be borrowed or something?)

**Comment by domenukk** [comment]
@addisoncrump just open a new PR if you manage to do it with 'static str somehow

**Pull request merged**